planner/core/rule_join_reorder.go

```diff
@@ -33,31 +33,64 @@ import (
 //
 // For example: "InnerJoin(InnerJoin(a, b), LeftJoin(c, d))"
 // results in a join group {a, b, c, d}.
-func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction,
-	otherConds []expression.Expression, joinTypes []*joinTypeWithExtMsg, hintInfo []*tableHintInfo, hasOuterJoin bool) {
+func extractJoinGroup(p LogicalPlan) *joinGroupResult {
+	joinMethodHintInfo := make(map[int]*joinMethodHint)
+	var (
+		group             []LogicalPlan
+		joinOrderHintInfo []*tableHintInfo
+		eqEdges           []*expression.ScalarFunction
+		otherConds        []expression.Expression
+		joinTypes         []*joinTypeWithExtMsg
+		hasOuterJoin      bool
+	)
 	join, isJoin := p.(*LogicalJoin)
 	if isJoin && join.preferJoinOrder {
 		// When there is a leading hint, the hint may not take effect for other reasons.
 		// For example, the join type is cross join or straight join, or exists the join algorithm hint, etc.
 		// We need to return the hint information to warn
-		hintInfo = append(hintInfo, join.hintInfo)
+		joinOrderHintInfo = append(joinOrderHintInfo, join.hintInfo)
+	}
+	// `leftHasHint` and `rightHasHint` are used to record whether the left child and right child are set by the join method hint.
+	leftHasHint, rightHasHint := false, false
+	if isJoin && p.SCtx().GetSessionVars().EnableAdvancedJoinHint && join.preferJoinType > uint(0) {
+		// If the current join node has the join method hint, we should store the hint information and restore it when we have finished the join reorder process.
+		if join.leftPreferJoinType > uint(0) {
+			joinMethodHintInfo[join.children[0].ID()] = &joinMethodHint{join.leftPreferJoinType, join.hintInfo}
```

Comment:
We didn't split the `join.hintInfo` to the left and the right side like the `join.preferJoinType`. Is it correct to put it here directly?

Reply:
Yes, you are right. I will put it after we truely split the join node.